Bluetooth: Drivers: HCI: Nordic improvements

rugeGerritsen opened the pull request:
Nordic devices are commonly used with the nRF Connect SDK.
There the SoftDevice Controller is set as the default
Bluetooth Controller. To avoid confusion when reading DTS
and Kconfig files, clarify this by adding a note.

jhedberg commented on dts/bindings/bluetooth/zephyr,bt-hci-virtual.yaml:
```diff
@@ -0,0 +1,11 @@
+description: Bluetooth HCI provided over a virtual bus
+
+compatible: "zephyr,bt-hci-virtual"
```
There are other HCI drivers providing access to a "local" controller, typically implemented as a binary blob (west blobs fetch ...), but still the same architecture as with the native controller or the SoftDevice one:
```shell
❯ git grep VIRTUAL dts/bindings/bluetooth/
dts/bindings/bluetooth/bt-hci.yaml:      - "BT_HCI_BUS_VIRTUAL"
dts/bindings/bluetooth/infineon,cat1-bless-hci.yaml:    default: "BT_HCI_BUS_VIRTUAL"
dts/bindings/bluetooth/silabs,bt-hci.yaml:    default: "BT_HCI_BUS_VIRTUAL"
dts/bindings/bluetooth/zephyr,bt-hci-ll-sw-split.yaml:    default: "BT_HCI_BUS_VIRTUAL"
```

In order not to cause confusion, I don't think we should have one of these bindings call itself generically "bt-hci-virtual". When using the SoftDevice controller, don't you anyway need a different set of Kconfig options enabled? If so, then having a dedicated binding for it would make sense as that could be linked to a separate KConfig option.

cvinayak replied:
I suggested the unique binding name, as the DT Kconfig option would select the corresponding Controller Kconfigs. 

> What we want to avoid is to overrides the device tree bindings in the zephyr repository or require adding an additional device tree property to all the socs.

Some of the SoCs/broads/applications would switch Controller depending on feature/code size, it is best to have separate binding name. (i see it like switching BT UART, console UART etc.)

jori-nordic replied:
> require adding an additional device tree property to all the socs

An option is to add a SDC dts overlay to all radio cores in NCS. In that one, we would override the `zephyr,bt-ll-sw-split` prop. 
I prefer to keep a one-to-one relationship between `compatible` string and `.c` driver source if possible.

jhedberg reviewed (commented):
Some comments regarding the binding naming. FWIW, the ll-sw-split name was originally suggested by @cvinayak

Even though you might be able to reuse the native controller binding with the SoftDevice controller, I think a clean approach would be to have a dedicated binding and compatible string for it (`nordic,...`).

rugeGerritsen commented on dts/common/nordic/nrf54l15.dtsi:
```diff
@@ -267,7 +267,7 @@
 				bt_hci_controller: bt_hci_controller {
 					compatible = "zephyr,bt-hci-virtual";
-					status = "disabled";
+					status = "okay";
```
I set the status to ok to match what we do for other Nordic SoCs. I noticed that https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/arm/nordic/nrf54l15_cpuapp.dtsi#L37 already sets the status to ok. So maybe this entry needs to be removed?

jhedberg commented on dts/arm/nordic/nrf51822.dtsi:
```diff
@@ -76,7 +76,7 @@
 
 			bt_hci_controller: bt_hci_controller {
-				compatible = "zephyr,bt-hci-ll-sw-split";
+				compatible = "nordic,bt-hci-virtual";
```
The binding still uses "zephyr,..." so I can't really see how this will work. Furthermore, since the binding is essentially for `subsys/bluetooth/controller/hci/hci_driver.c` that's (AFAIK) not Nordic specific, rather it can be used with any Lower Link Layer.

jori-nordic replied:
yes I agree: https://github.com/zephyrproject-rtos/zephyr/pull/74168#discussion_r1637861621

rugeGerritsen replied:
Yes. This was a temporary push. Discussing internally how to resolve this in a nice way. It would be nice if could avoid having two instances here

jori-nordic commented:
Discussed in Bluetooth WG:
- we can use an overlay downstream, applied in cmake. this overlay can just override the compatible string.
- this overlay could make use of HWMv2's board/soc revision functionality. This way we could have e.g. `nrf52840dk/nrf52840@bt-ll-sw` or `nrf52840dk/nrf52840@bt-sdc` to select the correct node

rugeGerritsen committed:
Bluetooth: Controller: Clarify that it can be overridden out of tree

Nordic devices are commonly used with the nRF Connect SDK.
There the SoftDevice Controller is set as the default
Bluetooth Controller. To avoid confusion when reading DTS
and Kconfig files, clarify this by adding a note.

Signed-off-by: Rubin Gerritsen <rubin.gerritsen@nordicsemi.no>
